common/replay/replay_guardian.go

```diff
@@ -0,0 +1,169 @@
+// Increasing this value permits more leniency in the timestamp of incoming requests, at the potential cost of a higher
+// memory overhead.
+func NewReplayGuardian(
+	timeSource func() time.Time,
+	maxTimeInPast time.Duration,
+	maxTimeInFuture time.Duration) ReplayGuardian {
+
+	return &replayGuardian{
+		timeSource:      timeSource,
+		maxTimeInFuture: maxTimeInFuture,
+		maxTimeInPast:   maxTimeInPast,
+		observedHashes:  make(map[string]struct{}),
+		expirationQueue: priorityqueue.NewWith(compareHashWithExpiration),
+	}
+}
+
+var _ ReplayGuardian = &replayGuardian{}
+
+// hashWithExpiration is a request hash with that request's expiration time.
+type hashWithExpiration struct {
+	hash       string
+	expiration time.Time
+}
+
+// compareKeyWithExpiration compares two hashWithExpiration objects by their expiration time. Used to create
+// a priority queue based on expiration time.
+func compareHashWithExpiration(a interface{}, b interface{}) int {
+
+	keyA := a.(*hashWithExpiration)
+	keyB := b.(*hashWithExpiration)
+
+	if keyA.expiration.Before(keyB.expiration) {
+		return -1
+	} else if keyA.expiration.After(keyB.expiration) {
+		return 1
+	}
+	return 0
+}
+
+// replayGuardian is an implementation of ReplayGuardian.
+type replayGuardian struct {
+	// The time source. In production use cases, this is likely to just be time.Now.
+	timeSource func() time.Time
+
+	// The maximum amount of time that a request's timestamp can be ahead of the local wall clock time.
+	maxTimeInFuture time.Duration
+
+	// The maximum amount of time that a request's timestamp can be behind the local wall clock time.
+	maxTimeInPast time.Duration
+
+	// A set of hashes that have been observed within the time window.
+	observedHashes map[string]struct{}
+
+	// A queue of observed hashes, ordered by expiration time. Used to prune old hashes.
+	expirationQueue *priorityqueue.Queue
+
+	// A mutex to protect the observedHashes and expirationQueue.
+	lock sync.Mutex
+}
+
+func (r *replayGuardian) VerifyRequest(requestHash []byte, requestTimestamp time.Time) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	now := r.timeSource()
+
+	// Do maintenance on the observedHashes set and expirationQueue.
+	r.pruneObservedHashes(now)
+
+	// Reject requests that fall outside the time window we are tracking.
+	err := r.verifyTimestamp(now, requestTimestamp)
+	if err != nil {
+		return err
+	}
+
+	// If we've reached this point, then the request will still be in the observedHashes set if it is a replay.
+	if _, ok := r.observedHashes[string(requestHash)]; ok {
+		return fmt.Errorf("request hash %x has been observed before", requestHash)
+	}
+
+	// The request is not a replay. Add the hash to the observedHashes set and the expirationQueue.
+	r.observedHashes[string(requestHash)] = struct{}{}
+	r.expirationQueue.Enqueue(&hashWithExpiration{
+		hash:       string(requestHash),
+		expiration: now.Add(r.maxTimeInPast),
```

Comment:
It seems like defining the expiration based on `maxTimeInPast` is kinda arbitrary? Couldn't it just as well be `maxTimeInFuture`, or `maxTimeInPast + maxTimeInFuture`? Or a separately configured value entirely? Why `maxTimeInPast`

Reply:
Fixed based on our discussion